### PR TITLE
[fix][client] Fix deadlock when sending chunked messages with BlockIFQueueFull enabled

### LINKED ISSUE #17446: [Bug] Producer may be permanently blocked by chunking messages when blockIfQueueFull is enabled
### Search before asking

- [X] I searched in the [issues](https://github.com/apache/pulsar/issues) and found nothing similar.

### Version

Pulsar 2.11

### Minimal reproduce step

It can be reproduced simply by following.
```
  @Test
  public void testMiniChunkingBlockIfQueueFull() throws Exception {
      this.conf.setMaxMessageSize(1000);

      final String topicName = "persistent://my-property/my-ns/testChunkingWithOrderingKey";

      final ExecutorService exec = Executors.newFixedThreadPool(1);

      @Cleanup
      Producer<byte[]> producer = pulsarClient.newProducer().topic(topicName).enableChunking(true)
              .chunkMaxMessageSize(1)
              .blockIfQueueFull(true)
              .maxPendingMessages(1000)
              .enableBatching(false).create();

      byte[] data = RandomUtils.nextBytes(1001);
      producer.newMessage().value(data).send();
  }
```

### What did you expect to see?

The reason for this bug is how the chunk message semaphore is obtained.
https://github.com/apache/pulsar/blob/359cfa7bc05775bf6dd004f21b9907610ed3b3d5/pulsar-client/src/main/java/org/apache/pulsar/client/impl/ProducerImpl.java#L520-L527
When a large message is split into a large number of chunks (i.e. the message is too big or the `chunkMaxMessageSize` is set too small), all the remaining semaphores will be acquired. The sending (send() and sendAsync()) of large message will be blocked by itself forever.

By the way, once blockIfQueueFull/maxPendingMessages/chunking are enabled at the same time, this risk of deadlock exists even if the number of chunks of a single message is not very large.

### What did you see instead?

//

### Anything else?

_No response_

### Are you willing to submit a PR?

- [X] I'm willing to submit a PR!

### PR DESCRIPTION
**Pull request by RobertIndie**
Fixes #17446

### Motivation

When the `isBlockIfQueueFull` is enabled for the producer, if the producer sends a very large message(# of total chunks > maxPendingMessages) or the producer sends too many large messages concurrently, there will be deadlock when the producer tries too acquire the permit before publishing these messages.

The root cause is that the producer will try to acquire all permits for all chunks before publishing them. It will be blocked if there are not enough permits. For example, if the payload size of the message is 55 MB which will be split into 11 chunks and the maxPendingMessages size is 10, it will be blocked forever. Even if the message is not too large(like 30MB) but sends 5 such messages concurrently, it may also be blocked forever.

### Modifications

* If the `BlockIfQueueFull` is enabled, the producer will only acquire a single permit for each chunk before publishing them. 

This PR only affects the case of `BlockIfQueueFull` enabled when sending chunked messages, and it will not affect other existing behaviors.

### Verifying this change

- [ ] Make sure that the change passes the CI checks.

This change added tests.

### Does this pull request potentially affect one of the following parts:

*If the box was checked, please highlight the changes*

- [ ] Dependencies (add or upgrade a dependency)
- [ ] The public API
- [ ] The schema
- [ ] The default values of configurations
- [ ] The binary protocol
- [ ] The REST endpoints
- [ ] The admin CLI options
- [ ] Anything that affects deployment

### Documentation

- [ ] `doc-required` 
(Your PR needs to update docs and you will update later)

- [x] `doc-not-needed` 
(Please explain why)

- [ ] `doc` 
(Your PR contains doc changes)

- [ ] `doc-complete`
(Docs have been already added)

### Matching PR in forked repository

PR in forked repository: https://github.com/RobertIndie/pulsar/pull/5

### CODE REVIEW

File: pulsar-broker/src/test/java/org/apache/pulsar/client/impl/MessageChunkingTest.java [hunk 1]
```diff
@@ -573,6 +574,54 @@ public void testChunkSize() throws Exception {
         }
     }
 
+    @Test
+    public void testBlockIfQueueFUllWhenChunking() throws Exception {
+        this.conf.setMaxMessageSize(50);
+
+        final Producer<String> producer = pulsarClient.newProducer(Schema.STRING)
```
Comment (codelipenghui) [inline pulsar-broker/src/test/java/org/apache/pulsar/client/impl/MessageChunkingTest.java:581]:
```suggestion
        @Cleanup
        final Producer<String> producer = pulsarClient.newProducer(Schema.STRING)
```

**Review by codelipenghui — APPROVED** [review]

File: pulsar-broker/src/test/java/org/apache/pulsar/client/impl/MessageChunkingTest.java [hunk 2]
```diff
@@ -573,6 +574,55 @@ public void testChunkSize() throws Exception {
         }
     }
 
+    @Test
+    public void testBlockIfQueueFUllWhenChunking() throws Exception {
```
Comment (BewareMyPower) [inline pulsar-broker/src/test/java/org/apache/pulsar/client/impl/MessageChunkingTest.java:578]:
```suggestion
    public void testBlockIfQueueFullWhenChunking() throws Exception {
```

A typo

File: pulsar-client/src/main/java/org/apache/pulsar/client/impl/ProducerImpl.java [hunk 3]
```diff
@@ -519,35 +519,42 @@ public void sendAsync(Message<?> message, SendCallback callback) {
...
+                        msg.getMessageBuilder().setOrderingKey(orderingKey);
                     }
+                }
+                if (chunkId > 0 && conf.isBlockIfQueueFull() && !canEnqueueRequest(callback,
+                        message.getSequenceId(), 0 /* The memory was already reserved */)) {
+                    client.getMemoryLimitController().releaseMemory(uncompressedSize - readStartIndex);
+                    semaphoreRelease(totalChunks - chunkId);
+                    return;
+                }
+                synchronized (this) {
```
[the remark this answers is not in the document]
Comment (BewareMyPower) [reply pulsar-client/src/main/java/org/apache/pulsar/client/impl/ProducerImpl.java:557]: I agree that the synchronized block should be shrinked. But could you explain in details about why will it cause another deadlock? With the previous synchronized block, the test still succeeded.

File: pulsar-client/src/main/java/org/apache/pulsar/client/impl/ProducerImpl.java [hunk 4]
```diff
@@ -519,35 +519,42 @@ public void sendAsync(Message<?> message, SendCallback callback) {
...
+                    if (orderingKey != null) {
+                        msg.getMessageBuilder().setOrderingKey(orderingKey);
                     }
+                }
+                if (chunkId > 0 && conf.isBlockIfQueueFull() && !canEnqueueRequest(callback,
+                        message.getSequenceId(), 0 /* The memory was already reserved */)) {
+                    client.getMemoryLimitController().releaseMemory(uncompressedSize - readStartIndex);
+                    semaphoreRelease(totalChunks - chunkId);
+                    return;
+                }
```
Comment (BewareMyPower) [inline pulsar-client/src/main/java/org/apache/pulsar/client/impl/ProducerImpl.java:556]: `canEnqueueRequest` always return true when `conf.isBlockIfQueueFull()` returns true. So it could never go into this if block.

**Review by BewareMyPower — CHANGES_REQUESTED** [review]

File: pulsar-client/src/main/java/org/apache/pulsar/client/impl/ProducerImpl.java [hunk 5]
```diff
@@ -519,35 +519,42 @@ public void sendAsync(Message<?> message, SendCallback callback) {
...
         // chunked message also sent individually so, try to acquire send-permits
         for (int i = 0; i < (totalChunks - 1); i++) {
-            if (!canEnqueueRequest(callback, message.getSequenceId(), 0 /* The memory was already reserved */)) {
+            if (!conf.isBlockIfQueueFull() && !canEnqueueRequest(callback, message.getSequenceId(),
+                    0 /* The memory was already reserved */)) {
                 client.getMemoryLimitController().releaseMemory(uncompressedSize);
                 semaphoreRelease(i + 1);
                 return;
             }
         }
```
Comment (BewareMyPower) [inline pulsar-client/src/main/java/org/apache/pulsar/client/impl/ProducerImpl.java:528]: The root cause is that for a chunked message with N chunks, since chunks are sent asynchronously one by one in the same thread, the time order will be:

1. Chunk 0 before send: Acquire 1 semaphore and `size0` memory.
2. Chunk 0 after send: Release 1 semaphore and `size0` memory.
3. Chunk 1 before send: Acquire 1 semaphore and `size1` memory.
4. Chunk 1 after send: Release 1 semaphore and `size1` memory. 
5. ...

In this case, we don't need N spots from the semaphore and `sum(size0, size1, ...)` bytes from the memory limiter. Instead, we only need 1 spot from the semaphore and `max(size0, size1, ...)` bytes from the memory limiter.

Therefore, we should not call `canEnqueueRequest` for chunk 1 and later, no matter if BlockIfQueueFUll is enabled. Instead, we should call `canEnqueueRequest` in each iteration in the loop.

**Review by BewareMyPower — CHANGES_REQUESTED** [review]

File: pulsar-client/src/main/java/org/apache/pulsar/client/impl/ProducerImpl.java (hunk 5 above)
Comment (codelipenghui) [reply pulsar-client/src/main/java/org/apache/pulsar/client/impl/ProducerImpl.java:528]: I think it's a great point to make the implementation simple and easy to understand.

File: pulsar-client/src/main/java/org/apache/pulsar/client/impl/ProducerImpl.java (hunk 4 above)
Comment (RobertIndie) [reply pulsar-client/src/main/java/org/apache/pulsar/client/impl/ProducerImpl.java:556]: It does not always return true in that case. We need to handle the exception case. https://github.com/RobertIndie/pulsar/blob/c6ebe063fc1c4088f5500ffd7ab5645c14bd581c/pulsar-client/src/main/java/org/apache/pulsar/client/impl/ProducerImpl.java#L937-L940

File: pulsar-broker/src/test/java/org/apache/pulsar/client/impl/MessageChunkingTest.java [hunk 6]
```diff
@@ -573,6 +574,55 @@ public void testChunkSize() throws Exception {
...
+        });
+        try {
+            sendMsg.get(5, TimeUnit.SECONDS);
+        } catch (TimeoutException e) {
+            Assert.fail("Deadlock detected when sending large message.");
+        }
+
+        // Test sending multiple large messages (For every message, totalChunks < maxPendingMessages) concurrently
+        // should not cause the deadlock.
+        List<CompletableFuture<Void>> sendMsgFutures = Lists.newArrayList();
```
[the remark this answers is not in the document]
Comment (RobertIndie) [reply pulsar-broker/src/test/java/org/apache/pulsar/client/impl/MessageChunkingTest.java:608]: It's already imported. The compilation was fine in my repo: https://github.com/RobertIndie/pulsar/pull/5 I am investigating the CI problem.

[the remark this answers is not in the document]
Comment (BewareMyPower) [reply pulsar-broker/src/test/java/org/apache/pulsar/client/impl/MessageChunkingTest.java:608]: Could you point out in which line it's imported? I cannot find any `import xx.xx.Lists` in this file.

[the remark this answers is not in the document]
Comment (BewareMyPower) [reply pulsar-broker/src/test/java/org/apache/pulsar/client/impl/MessageChunkingTest.java:608]: image

It also failed in my Intellij Idea on Windows.

File: pulsar-client/src/main/java/org/apache/pulsar/client/impl/ProducerImpl.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (BewareMyPower) [reply pulsar-client/src/main/java/org/apache/pulsar/client/impl/ProducerImpl.java:557]: Just see your branch is far behind the master, please rebase to latest master.

File: pulsar-client/src/main/java/org/apache/pulsar/client/impl/ProducerImpl.java (hunk 5 above)
Comment (RobertIndie) [reply pulsar-client/src/main/java/org/apache/pulsar/client/impl/ProducerImpl.java:528]: @BewareMyPower Yes. Chunks are sent asynchronously one by one in the same thread. But from your example, it seems that they are sent synchronously one by one. My understanding is that there may be a case like this in your approach:

Chunk 0 before send: Acquire 1 semaphore and max(size...) memory
Chunk 1 before send: 
// Here we already occupy 2 semaphores and (size0+size1) memory. It's not the correct behavior.
Chunk 0 after send: 
Chunk 1 after send: Release 1 semaphore and max(size...) memory.
...

Please correct if I'm wrong.

File: pulsar-client/src/main/java/org/apache/pulsar/client/impl/ProducerImpl.java (hunk 4 above)
Comment (BewareMyPower) [reply pulsar-client/src/main/java/org/apache/pulsar/client/impl/ProducerImpl.java:556]: I see now. However, when an `InterruptedException` is thrown, it means `Semaphore#acquire` or `MemoryLimitController#reserveMemory` is interrupted.
- If `Semaphore` is interrupted, we should not release both.
- If `MemoryLimitController` is interrupted, we should only release the semaphore.

BTW, I think it's a bad design to just return false in this case for the different cases.

Comment (BewareMyPower) [reply pulsar-client/src/main/java/org/apache/pulsar/client/impl/ProducerImpl.java:556]: Anyway, it's just a corner case. It's not much related to this PR, I will push a new PR for it.

File: pulsar-broker/src/test/java/org/apache/pulsar/client/impl/MessageChunkingTest.java (hunk 6 above)
[the remark this answers is not in the document]
Comment (RobertIndie) [reply pulsar-broker/src/test/java/org/apache/pulsar/client/impl/MessageChunkingTest.java:608]: It's imported here: https://github.com/RobertIndie/pulsar/blob/c6ebe063fc1c4088f5500ffd7ab5645c14bd581c/pulsar-broker/src/test/java/org/apache/pulsar/client/impl/MessageChunkingTest.java#L26

Wired. I will rebase the branch.

File: pulsar-client/src/main/java/org/apache/pulsar/client/impl/ProducerImpl.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (RobertIndie) [reply pulsar-client/src/main/java/org/apache/pulsar/client/impl/ProducerImpl.java:557]: If the thread is blocked at `canEnqueueRequest`, the IO thread will try to acquire this lock after receiving the ack of the send request. And this will be a deadlock [here]()https://github.com/RobertIndie/pulsar/blob/c6ebe063fc1c4088f5500ffd7ab5645c14bd581c/pulsar-client/src/main/java/org/apache/pulsar/client/impl/ProducerImpl.java#L1126.

I can reproduce the deadlock after rebasing to the latest master branch.

File: pulsar-broker/src/test/java/org/apache/pulsar/client/impl/MessageChunkingTest.java (hunk 6 above)
[the remark this answers is not in the document]
Comment (RobertIndie) [reply pulsar-broker/src/test/java/org/apache/pulsar/client/impl/MessageChunkingTest.java:609]: Fixed it.

File: pulsar-client/src/main/java/org/apache/pulsar/client/impl/ProducerImpl.java (hunk 5 above)
Comment (BewareMyPower) [reply pulsar-client/src/main/java/org/apache/pulsar/client/impl/ProducerImpl.java:526]: Oh I see. Mark it as resolved.

File: pulsar-client/src/main/java/org/apache/pulsar/client/impl/ProducerImpl.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (BewareMyPower) [reply pulsar-client/src/main/java/org/apache/pulsar/client/impl/ProducerImpl.java:555]: I see. Yeah, it should be careful to include a blocking operation in a synchronized block.

**Review by BewareMyPower — APPROVED** [review]

File: pulsar-broker/src/test/java/org/apache/pulsar/client/impl/MessageChunkingTest.java (hunk 6 above)
[the remark this answers is not in the document]
Comment (BewareMyPower) [reply pulsar-broker/src/test/java/org/apache/pulsar/client/impl/MessageChunkingTest.java:609]:
```
[INFO] --- modernizer-maven-plugin:2.3.0:modernizer (modernizer) @ pulsar-broker ---
Error:  /home/runner/work/pulsar/pulsar/pulsar-broker/src/test/java/org/apache/pulsar/client/impl/MessageChunkingTest.java:609: Prefer java.util.ArrayList<>()
```

It looks like the modernizer-maven-plugin prefers using `ArrayList` directly.
